### PR TITLE
implement command teleporting to a player's last known logout location

### PR DESCRIPTION
**Pull request by ghost**
This PR contributes to (and closes) https://github.com/EssentialsX/Essentials/issues/1561. This implements a command that teleports you to a player's last known logout location.

### CODE REVIEW

**Comment by Ichbinjoe** [comment]
So when talking about this issue within a group of contributors, we thought it would be best if this were simply an add-on to the /tp (and friends) command. The functionality is similar enough to /tp that I don't see a reason why we need a new command just to handle offline player teleporting.

**Comment by mdcfe** [comment]
> So when talking about this issue within a group of contributors, we thought it would be best if this were simply an add-on to the /tp (and friends) command.

We did? I don't recall this consensus.

**Comment by Ichbinjoe** [comment]
So turns out I had those conversations in private with people who weren't md678585, so thats my mistake. 

Regardless, my opinion on the matter still stands. The functionality of teleportation to a player who is offline is very similar (logically to a user) to teleporting to a player who is online. I think this functionality should be combined with /tp and friends gated by a permission node.

**Comment by ghost** [comment]
> So turns out I had those conversations in private with people who weren't md678585, so thats my mistake.
> 
> Regardless, my opinion on the matter still stands. The functionality of teleportation to a player who is offline is very similar (logically to a user) to teleporting to a player who is online. I think this functionality should be combined with /tp and friends gated by a permission node.

Why should we modify `/tp`'s default behavior? I'm personally all for a separate command, to prevent confusion or player mismatches. We could send a message if someone tries to teleport to an offline player using the regular `/tp` command, assuming they have `essentials.offlinetp`, letting them know they can use the command to teleport to their last known logout location.

**Comment by mdcfe** [comment]
@Ichbinjoe 
> The functionality of teleportation to a player who is offline is very similar (logically to a user) to teleporting to a player who is online.

The functionality of requesting to teleport to someone online (`/tpa`) or teleporting while overriding someone's teleport toggle (`/tpo`) is also similar, yet they have their own commands. They're all teleportation, but they are distinct variations on `/tp`'s behaviour.

If a user runs `/tp <name>`, it teleports them to the first user who is online **whose username or displayname partially or fully matches `<name>`**.

Other commands in EssentialsX that support both offline and online players now generally target the **exact offline player** *before* partial matches of online players. Adding offline teleportation to `/tp`, if consistent with other commands, will prevent people using partial matches of online players if there is an offline player with that name. To players, moderators and likely even admins, this is a completely unexpected breaking change: for example, there's a player online with the nickname "MrTrent", but `/tp Trent` will teleport to an offline player called "Trent" for no apparent reason.

> I think this functionality should be combined with /tp and friends gated by a permission node.

Having a separate permission checked within `/tp` also causes problems. If we skip the offline player lookup when players don't have a permission, it then becomes hard to tell whether you don't have access to offline teleportation or you just got the username wrong. If we *don't* skip the player lookup and instead throw an error if it matches, it obstructs users who weren't trying to teleport to offline players. Separating this into its own command (like `/tpa[here]` and `/tpo[here]`) makes it plainly evident what the command you're running is supposed to achieve and clarifies why it may not have worked.

@latiku
> We could send a message if someone tries to teleport to an offline player using the regular `/tp` command, assuming they have `essentials.offlinetp`, letting them know they can use the command to teleport to their last known logout location.

I seem to recall suggesting this before - this is preferable as it is the least intrusive option. If you don't have permission, the command fails like every other permission and you're not nagged about features you're not allowed to use anyway.

**Comment by ghost** [comment]
Any suggestions as to what the default message to be for the message given when a user tries to use `/tp` to teleport to an offline player?

File: Essentials/src/com/earth2me/essentials/commands/Commandtp.java [hunk 1]
```diff
@@ -26,10 +28,16 @@ public void run(final Server server, final User user, final String commandLabel,
                 throw new NotEnoughArgumentsException();
 
             case 1:
-                final User player = getPlayer(server, user, args, 0);
+                final User player = getPlayer(server, user, args, 0, true, true);
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandtp.java]: `getHidden` should be false - afaik this currently lets people teleport to hidden players.

**Review by mdcfe — DISMISSED** [review]

File: Essentials/src/com/earth2me/essentials/commands/Commandtp.java (hunk 1 above)
Comment (ghost) [reply Essentials/src/com/earth2me/essentials/commands/Commandtp.java]: > getHidden should be false - afaik this currently lets people teleport to hidden players.

Sorry, haven't been on in a while. I'll fix this right now.

**Comment by pop4959** [comment]
This is a nitpick, but I think the command should be named /tpoffline instead of /offlinetp, since there is a precedent in Essentials for tp commands always starting with "tp". Examples below.

image

Otherwise, at a glance, this PR looks decent to me.

**Comment by ghost** [comment]
> This is a nitpick, but I think the command should be named /tpoffline instead of /offlinetp, since there is a precedent in Essentials for tp commands always starting with "tp". Examples below.
> 
> image
> 
> Otherwise, at a glance, this PR looks decent to me.

sure

File: Essentials/src/com/earth2me/essentials/commands/Commandtpoffline.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+public class Commandtpoffline extends EssentialsCommand {
+
+    public Commandtpoffline() {
+        super("tpoffline");
+    }
+
+    @Override
+    public void run(final Server server, final User user, final String label, final String[] args) throws Exception {
+        switch (args.length) {
```
Comment (Ichbinjoe) [inline Essentials/src/com/earth2me/essentials/commands/Commandtpoffline.java:18]: Nit: Why not if?

**Review by Ichbinjoe — APPROVED** [review]

**Pull request merged**